compiler/grainlsp/utils.re

```diff
@@ -0,0 +1,816 @@
+open Grain;
+open Compile;
+open Grain_parsing;
+open Grain_utils;
+open Grain_typed;
+open Grain_diagnostics;
+
+//  CompletionItemKind
+let item_kind_completion_text = 1;
+let item_kind_completion_function = 3;
+let item_kind_completion_constructor = 4;
+let item_kind_completion_variable = 6;
+let item_kind_completion_struct = 22;
+
+let markdown_grain = code => "```grain\n" ++ code ++ "\n```";
+
+type node_t =
+  | Expression(Typedtree.expression)
+  | Pattern(Typedtree.pattern)
+  | NotInRange
+  | Error(string);
+
+type source_text =
+  | Nothing
+  | Lident(string);
+
+let rec print_path = (ident: Path.t) => {
+  switch (ident) {
+  | PIdent(name) => name.name
+  | PExternal(externalIdent, second, _) =>
+    print_path(externalIdent) ++ "." ++ second
+  };
+};
+
+let comma_join = (~sep, ~print_fn, vals) => {
+  List.fold_left(
+    (acc, arg) => {
+      let existing =
+        if (acc == "") {
+          acc;
+        } else {
+          acc ++ sep;
+        };
+      existing ++ print_fn(arg);
+    },
+    "",
+    vals,
+  );
+};
+
+let loc_to_range = (pos: Location.t): Rpc.range_t => {
+  let (_, startline, startchar, _) =
+    Locations.get_raw_pos_info(pos.loc_start);
+  let (_, endline, endchar) =
+    Grain_parsing.Location.get_pos_info(pos.loc_end);
+
+  let range: Rpc.range_t = {
+    start_line: startline,
+    start_char: startchar,
+    end_line: endline,
+    end_char: endchar,
+  };
+  range;
+};
+
+let is_point_inside_stmt = (~line: int, loc: Grain_parsing.Location.t) => {
+  let (_, raw1l, raw1c, _) = Locations.get_raw_pos_info(loc.loc_start);
+  let (_, raw1le, raw1ce, _) = Locations.get_raw_pos_info(loc.loc_end);
+  if (line == raw1l || line == raw1le) {
+    true;
+  } else {
+    line > raw1l && line < raw1le;
+  };
+};
+let is_point_inside_location =
+    (~line: int, ~char: int, loc1: Grain_parsing.Location.t) => {
+  let (_, raw1l, raw1c, _) = Locations.get_raw_pos_info(loc1.loc_start);
+  let (_, raw1le, raw1ce, _) = Locations.get_raw_pos_info(loc1.loc_end);
+
+  if (line == raw1l) {
+    if (char >= raw1c) {
+      if (line == raw1le) {
+        char <= raw1ce;
+      } else {
+        true;
+      };
+    } else {
+      false;
+    };
```

Comment:
lol, I'll trust you

Reply:
I know it's ugly, but it's more explicit than trying to use and and or operators